File: compositionlib/src/main/scala/optimisation/consonancemetrics/Harmonicity.scala
package optimisation.consonancemetrics

import breeze.linalg.{Axis, DenseMatrix, DenseVector, sum}
import breeze.stats.distributions.{Gaussian, RandBasis}
import midi.FLStudioMIDIMap
import models.Primitives.PitchName

object Harmonicity {
  val log2 = (x: Double) => math.log10(x) / math.log10(2.0)

  val num_harmonics = 12
  val rho = 0.75
  val sigma = 6.83
  val array_dim = 1200

  def getComplexTone(pc: Int, weight: Double) = {
    val pcs = (1 to num_harmonics).map(i => {
      ((pc * array_dim / 12) + (array_dim * log2(i))) % array_dim
    })

    val weights = (1 to num_harmonics).map(i => {
      weight / math.pow(i, rho)
    })

    val spectra = (pcs zip weights).map(x => {
      val (pc, weight) = x
      pc_spectrum_template_2(array_dim, pc, weight, sigma).toArray
    }).toArray

    val m = new DenseMatrix(array_dim, spectra.length, spectra.flatten)

    sum(m, Axis._1)

  }

  def pc_spectrum_template_2(array_dim: Int, mean: Double, mass: Double, sigma: Double, truncation_point: Int = 12) = {
    val origin = math.round(mean)
    val template = pc_spectrum_template_1(array_dim, sigma, truncation_point)
    val scaled = template.map(_ * mass)
    val indices = (0 until array_dim).map(x => {
      val x_minus_origin = x - origin
      val i = if (x_minus_origin >= 0) x_minus_origin % array_dim else (array_dim - origin + x) % array_dim
      i
    })
    val output = indices.map(i => scaled(i.toInt))
    //output <- scaled[((seq - origin) %% array_dim) + 1]
    output
  }

  def pc_spectrum_template_1(array_dim: Int, sigma: Double, truncation_point: Int) = {
    val gaussian = Gaussian(0, 6.83)(RandBasis.mt0)
    val limit = math.floor(sigma * 12).toInt
    val template = (0 until array_dim).map(i => {
      i match {
        case x if x <= limit => gaussian.pdf(i)
        case x if x >= array_dim - limit => gaussian.pdf(array_dim - x)
        case _ => 0.0
      }
    })
    template
  }

  def get_milne_pc_spectrum(chord: Set[Int], weights: Array[Double]) = {
    val spectra = (chord.toArray zip weights).map(x => getComplexTone(x._1, x._2))
    val milne_pc_spectrum = sum(DenseMatrix(spectra: _*), Axis._0)
    milne_pc_spectrum
  }

  def sweep_template(x: Vector[Double], template: DenseVector[Double]) = {
    (0 until x.length).map(i => {
      val (first, last) = x.splitAt(i)
      1 - breeze.linalg.cosineDistance(template, DenseVector((last ++ first).toArray))
    })
  }

  def getHarmonicity(chord: List[Int]) = {
    val x = chord.map(i => i % 12).toSet
    val weights = Array.fill(x.size)(1.0)

    val milne_pc_spectrum = get_milne_pc_spectrum(x, weights)
    val template = get_milne_pc_spectrum(Set(0), Array(1.0))
    val y = sweep_template(milne_pc_spectrum.inner.toScalaVector, template.t)
    val sum_y = y.sum
    val probs = y.map(x => x / sum_y)
    val uniform_probs = 1.0 / probs.size
    val nonzero_probs = probs.filter(_ > 0.0)
    val harmonicity = nonzero_probs.map(p => p * log2(p / uniform_probs)).sum
    harmonicity
  }

  def getHarmonicityOfPitchNames(pitchNames: List[PitchName]) = {
    getHarmonicity(pitchNames.map(name => FLStudioMIDIMap.midiMap.get(name).get))
  }

}

File: HarmonyFunctions/src/data_generators/Harmonicity.py
import tensorflow as tf
import numpy as np
import math
from scipy.stats import norm

graph = tf.Graph()

array_dim = 1200
num_harmonics = 12
array_dim_float = tf.constant(1200, tf.float32)
num_harmonics_float = tf.constant(12, tf.float32)
sigma = 6.83
rho = 0.75

def log2tf(x):
    return tf.math.divide(tf.math.log(x), tf.math.log(2.0))

def get_pc_spectrum_template_1(array_dim, sigma): 
    limit = math.floor(sigma * 12) + 1
    template_values = np.concatenate([
      norm.pdf(np.arange(limit), 0, sigma), 
      np.zeros(array_dim - 2*limit, dtype=np.int32), 
      norm.pdf(array_dim - np.arange(array_dim - limit, array_dim), 0, sigma)
      ])

    return tf.constant(template_values, dtype=tf.float32)

pc_spectrum_template_1 = get_pc_spectrum_template_1(array_dim, sigma)


def get_pc_spectrum_template_2(mean, mass, chordSize):

    scaled = tf.math.multiply(
        tf.reshape(pc_spectrum_template_1, (1,1,-1)),
        tf.expand_dims(mass, 2)
    )

    indexRange = tf.range(0, array_dim, 1)

    origin = tf.cast(tf.math.round(mean), tf.int32)

    x_minus_origin = tf.math.subtract(
        tf.reshape(indexRange, (1,1,-1)),
        tf.expand_dims(origin, 2)
    )
    indices = tf.math.floormod(x_minus_origin, array_dim)

    spectralBandIndices = tf.reshape(
        indices,
        (-1))

    harmonicIndices = tf.tile(
        tf.reshape(
        tf.linalg.matrix_transpose(
            tf.reshape(
            tf.tile(
                tf.range(0, num_harmonics, 1),
                np.array([array_dim])
            ), (array_dim,-1)
            )
        ),
        -1
        ),
        tf.expand_dims(chordSize, 0)
    )

    chordIndices = tf.reshape(
        tf.linalg.matrix_transpose(
        tf.reshape(
            tf.tile(
            tf.range(0, chordSize, 1),
            np.array([array_dim*num_harmonics])
            ),
            (array_dim*num_harmonics, -1)
        )
        ),
        -1
    )

    gatherIndices = tf.linalg.matrix_transpose(
        tf.stack([chordIndices, harmonicIndices, spectralBandIndices])
    )

    gathered = tf.reshape(
        tf.gather_nd(scaled, gatherIndices),
        tf.concat([tf.reshape(chordSize, 1), (num_harmonics,array_dim)], 0)
    )

    return gathered
    

def getComplexTone(tone, weight):
    harmonicIndices = tf.math.add(tf.range(0, num_harmonics, 1), 1)
    harmonicIndicesDouble = tf.cast(harmonicIndices, tf.float32)
    dimMulIndex = tf.math.multiply(array_dim, log2tf(harmonicIndicesDouble))

    toneAsDouble = tf.cast(tone, tf.float32)
    dimMulTone = tf.math.multiply(toneAsDouble, tf.math.divide(array_dim_float, num_harmonics_float))

    pcs = tf.math.mod(
      tf.math.add(
        tf.expand_dims(dimMulTone, 1),
        tf.expand_dims(dimMulIndex, 0)
      ), array_dim)

    weights = tf.math.divide(
      tf.expand_dims(weight, 1),
      tf.expand_dims(tf.math.pow(harmonicIndicesDouble, rho), 0)
    )

    pc_spectrum_template_2 = get_pc_spectrum_template_2(pcs, weights, tf.size(tone))
    return tf.reduce_sum(pc_spectrum_template_2, 1)
  



def get_milne_pc_spectrum(chord, weights):
    spectra = getComplexTone(chord, weights)
    milne_pc_spectrum = tf.reduce_sum(spectra, 0)
    return milne_pc_spectrum
  

def sweep_template(milne_pc_spectrum, template):
    indices = tf.range(0, array_dim, 1)

    colIndices = tf.reshape(
      tf.math.mod(
        tf.math.add(
          tf.reshape(
            tf.tile(
              indices,
              np.array([array_dim])
            ),
            (array_dim,-1)
          ),
          tf.expand_dims(indices, 1)
        ),
        array_dim
      ),
      -1
    )

    rowIndices = tf.reshape(
      tf.linalg.matrix_transpose(
        tf.reshape(
          tf.tile(indices, np.array([array_dim])),
          (array_dim,-1)
        )
      ),
      -1
    )

    matrixTemplate = tf.reshape(
        tf.tile(milne_pc_spectrum, np.array([array_dim])),
        (array_dim,-1)
      ) # np.sum(matrixTemplate) 9792.714; matrixTemplate[0] [0.12607166, 0.12472759, 0.12078077, ..., 0.11447829, 0.12078077, 0.12472759]; 
      #np.sum(matrixTemplate, axis=0) [151.28366, 149.6719 , 144.93605, ..., 137.37413, 144.93605, 149.6719 ]

    matrix = tf.reshape(
      tf.gather_nd(
        matrixTemplate,
        tf.linalg.matrix_transpose(
          tf.stack([rowIndices, colIndices])
        )
      ),
      (array_dim,array_dim)
    ) 

    matrixNorm = tf.math.pow(
      tf.linalg.einsum("ij,ij->i", matrix, matrix),
      0.5
    )

    vectorNorm = tf.math.pow(
      tf.linalg.einsum("i,i->", template, template),
      0.5
    )

    dot = tf.linalg.einsum("ij,i->j", matrix, template)

    cosineSimilarity = tf.math.divide(
      dot,
      tf.math.multiply(
        matrixNorm,
        vectorNorm
      )
    )

    return cosineSimilarity
  

def calcHarmonicity(chord, weights):
    #chord = np.array([0,7], dtype = np.int32)
    #weights = np.array([1.0, 1.0], np.float32)

    milne_pc_spectrum = get_milne_pc_spectrum(chord, weights) #[0.12607166, 0.12472759, 0.12078077, ..., 0.11447829, 0.12078077, 0.12472759] 8.160593
    template = get_milne_pc_spectrum(tf.constant([0], dtype=tf.float32), tf.constant([1.0], dtype=tf.float32)) #4.0802965
    y = sweep_template(milne_pc_spectrum, template)
    sum_y = tf.math.reduce_sum(y, 0)
    probs = tf.math.divide(y, sum_y) #[1200]

    uniform_probs = tf.math.divide(
      tf.constant(1.0),
      tf.cast(tf.size(probs), tf.float32)
    ) #[]
    mask = tf.math.greater(probs, tf.constant(0.0)) #[1200]
    nonzero_probs = tf.boolean_mask(probs, mask) #[1200]
    harmonicity = tf.math.reduce_sum(tf.math.multiply(
      nonzero_probs,
      log2tf(tf.math.divide(nonzero_probs, uniform_probs))
    ), 0) #[]

    return tf.expand_dims(harmonicity, 0)


def getGraph():
    chordPlaceHolder = tf.placeholder(tf.int32)
    weightsPlaceHolder = tf.placeholder(tf.float32)
    harmonicity = calcHarmonicity(chordPlaceHolder, weightsPlaceHolder)
    return (graph, harmonicity, chordPlaceHolder, weightsPlaceHolder)
